sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3628,9 +3648,16 @@ pi_result piMemRelease(pi_mem Mem) {
             ZeUSMImport.doZeUSMRelease(driverHandle, Mem->MapHostPtr);
           } else {
             if (enableBufferPooling()) {
-              PI_CALL(piextUSMFree(Mem->Context, Mem->getZeHandle()));
+              pi_platform Plt = Mem->Context->Devices[0]->Platform;
+              std::unique_lock<std::mutex> ContextsLock(Plt->ContextsMutex,
+                                                        std::defer_lock);
+              if (IndirectAccessTrackingEnabled)
+                ContextsLock.lock();
```

Comment:
Why does it need to be here and not inside the `USMFreeHelper` like it is inside of `ZeMemFreeHelper`? BTW, can we always call `USMFreeHelper`?

Reply:
There are several place where need to lock this mutex beforehand - before call to USMFreeHelper, so can't lock inside this function because it will cause deadlock.
USMFreeHelper performs usm pooling and zeMemFreeHelper doesn't perform usm pooling, so can't just replace one call with another without refactoring.

I agree both concerns most likely can be addressed with refactoring but changes are not trivial and  not not related to this PR, so I would prefer to do this separately.

Reply:
> USMFreeHelper performs usm pooling and zeMemFreeHelper doesn't perform usm pooling, so can't just replace one call with another without refactoring.

And we want everything to use "pooling", but I agree it better go in separately.